firedrake/slate/slac/kernel_builder.py

```diff
@@ -1,76 +1,70 @@
-
-from collections import OrderedDict
+from abc import ABCMeta, abstractmethod, abstractproperty
 
 from coffee import base as ast
 
-from firedrake.slate.slate import (TensorBase, Tensor, TensorOp,
-                                   Action, Inverse)
-from firedrake.slate.slac.utils import (Transformer, traverse_dags,
-                                        collect_reference_count)
+from collections import OrderedDict
+
+from firedrake.slate.slate import (TensorBase, Tensor,
+                                   TensorOp, Action,
+                                   Negative, Transpose)
+from firedrake.slate.slac.utils import (traverse_dags,
+                                        collect_reference_count,
+                                        count_operands)
 from firedrake.utils import cached_property
 
 from ufl import MixedElement
 
 
-class KernelBuilder(object):
-    """A helper class for constructing Slate kernels.
-
-    This class provides access to all temporaries and subkernels associated
-    with a Slate expression. If the Slate expression contains nodes that
-    require operations on already assembled data (such as the action of a
-    slate tensor on a `ufl.Coefficient`), this class provides access to the
-    expression which needs special handling.
+class KernelBuilderBase(object, metaclass=ABCMeta):
+    """A base helper class for constructing Slate kernels."""
 
-    Instructions for assembling the full kernel AST of a Slate expression is
-    provided by the method `construct_ast`.
-    """
     def __init__(self, expression, tsfc_parameters=None):
-        """Constructor for the KernelBuilder class.
+        """Constructor for the KernelBuilderBase class.
 
         :arg expression: a :class:`TensorBase` object.
         :arg tsfc_parameters: an optional `dict` of parameters to provide to
                               TSFC when constructing subkernels associated
                               with the expression.
         """
         assert isinstance(expression, TensorBase)
+
+        # Collect reference counts
+        ref_counts = collect_reference_count([expression])
```

Comment:
No need for a comment like this, if the code on the following line doesn't collect reference counts then it is highly misnamed.